File: customview/src/main/java/dinson/customview/_globle/CrashHandler.java
package dinson.customview._globle;

import android.content.Context;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.List;

import dinson.customview.utils.LogUtils;

/**
 * 异常拦截器
 */
public class CrashHandler implements Thread.UncaughtExceptionHandler {

    private static CrashHandler instance;
    private Context context;
    private String errorLogDir = null;
    private SimpleDateFormat sdf = new SimpleDateFormat("yyyyMMdd_HHmmss");
    private final int expirationDate = 7;//设置过期时间(单位：天)

    private CrashHandler() {
    }

    public synchronized static CrashHandler getInstance() {
        if (instance == null) {
            instance = new CrashHandler();
        }
        return instance;
    }

    public void init(Context context, String errorLogDir) {
        this.context = context;
        this.errorLogDir = errorLogDir;
        Thread.setDefaultUncaughtExceptionHandler(this);
        //删除过期文件
        new Thread(new Runnable() {
            @Override
            public void run() {
                deleteExpirationFile();
            }
        }).start();
    }

    @Override
    public void uncaughtException(Thread thread, Throwable ex) {
        if (ex != null) {
            disposeThrowable(ex);
            ex.printStackTrace();
        }
    }

    private void disposeThrowable(Throwable ex) {
//        StringBuffer sb = new StringBuffer();
//        sb.append(ex.toString() + "\n");
//        StackTraceElement[] steArray = ex.getStackTrace();
//        for (StackTraceElement ste : steArray) {
//            sb.append("System.err at " + ste.getClassName() + "." + ste.getMethodName() + "(" + ste.getFileName() + ":" + ste.getLineNumber() + ")\n");
//        }
        if (errorLogDir != null) {
            try {
                File file = new File(errorLogDir + "error_" + sdf.format(System.currentTimeMillis()) + ".txt");
                if (!file.exists()) {
                    file.getParentFile().mkdirs();
                    file.createNewFile();
                }
//                FileOutputStream fos = new FileOutputStream(file);
//                fos.write(sb.toString().getBytes("utf-8"));
//                fos.flush();
//                fos.close();
//                fos = null;
                PrintWriter pw = new PrintWriter(new BufferedWriter(new FileWriter(file, true)));
                // 导出异常的调用栈信息
                ex.printStackTrace(pw);
                pw.println();
                pw.close();
            } catch (FileNotFoundException e) {

            } catch (IOException e) {
                e.printStackTrace();
            }

            LogUtils.e(ex.toString());

        }
        //退出程序
        android.os.Process.killProcess(android.os.Process.myPid());
        System.exit(1);
    }

    /**
     * 删除过期文件
     */
    private void deleteExpirationFile() {
        if (errorLogDir == null) {
            return;
        }
        long expirationTime = expirationDate * 24 * 3600 * 1000;
        long nowTime = System.currentTimeMillis();
        List<File> deleteList = new ArrayList<File>();
        File dir = new File(errorLogDir);
        if (dir.isDirectory()) {
            File[] files = dir.listFiles();
            for (File f : files) {
                if (f.exists()) {
                    long time = f.lastModified();
                    if (nowTime - time >= expirationTime) {
                        deleteList.add(f);
                    }
                }
            }
            for (File f : deleteList) {
                f.delete();
            }
        }
    }
}

File: blingbling/src/main/java/dinson/customview/_global/CrashHandler.kt
package dinson.customview._global

import dinson.customview.kotlin.loge
import java.io.BufferedWriter
import java.io.File
import java.io.FileWriter
import java.io.PrintWriter
import java.text.SimpleDateFormat
import java.util.*
import kotlin.system.exitProcess

/**
 * 异常拦截类
 */
object CrashHandler : Thread.UncaughtExceptionHandler {

    private lateinit var mErrorLogDir: String
    private const val mExpirationDate = 3   //文件的存活周期：单位：天
    private const val mPrefix = "error_"

    fun init(errorLogDir: String) {
        mErrorLogDir = errorLogDir
        Thread.setDefaultUncaughtExceptionHandler(this)
        //删除过期文件
        Thread(Runnable { deleteExpirationFile() }).start()
    }

    /**
     * 拦截的线程
     */
    override fun uncaughtException(p0: Thread?, ex: Throwable?) {
        ex?.let { disposeThrowable(ex) }
    }

    /**
     * 写文件
     */
    private fun disposeThrowable(ex: Throwable) {
        try {
            val nowTime = SimpleDateFormat("MMdd_HHmmss", Locale.CHINA).format(System.currentTimeMillis())
            val file = File("$mErrorLogDir$mPrefix$nowTime.txt")
            if (!file.exists()) {
                file.parentFile.mkdirs()
                file.createNewFile()
            }
            val pw = PrintWriter(BufferedWriter(FileWriter(file, true)))
            ex.printStackTrace(pw)
            pw.println()
            pw.close()
        } catch (e: Exception) {
            e.printStackTrace()
        }

        loge { ex.toString() }
        //退出程序
        android.os.Process.killProcess(android.os.Process.myPid())
        exitProcess(1)
    }


    /**
     * 删除过期文件
     */
    private fun deleteExpirationFile() {
        val expirationTime = (mExpirationDate * 24 * 3600 * 1000).toLong()
        val nowTime = System.currentTimeMillis()
        val dir = File(mErrorLogDir)
        if (dir.isDirectory) {
            dir.listFiles().filter { nowTime - it.lastModified() >= expirationTime }
                .forEach { it.delete() }
        }
    }
}